zipl/man/zipl.8
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH ZIPL 8 "Feb 2012" "s390-tools"
.SH NAME
zipl \- boot loader for IBM S/390 and zSeries architectures

.SH SYNOPSIS
.B zipl
[OPTION] [SECTION]

.SH DESCRIPTION
.B zipl
is a boot loader tool for IBM S/390 and zSeries machines. It can be used to
prepare devices for initial program load (IPL). The following functions are
supported:
.IP "     -"
booting a Linux kernel with optional ramdisk and kernel command line
.IP "     -"
taking a snapshot of the current system status (system dump)
.IP "     -"
loading a data file to initialize named saved segments (NSS)
.PP
Each of these operations is characterized by a boot configuration, i.e. a
set of required parameters.
.B zipl
supports three ways of specifying a boot configuration:
.IP "     -"
.B command line:
all parameters are provided through the command line switches described below.
Using this mode, only a single boot configuration can be specified.
.IP "     -"
.B configuration file:
parameters are provided by sections defined in a configuration file (see
.BR zipl.conf (5)).
Using a configuration file, you can either specify a single boot configuration
or a menu, i.e. a list of configurations from which users can choose at boot
time.
.IP "     -"
.B BLS config files:
boot configurations are specified using BootLoaderSpec (BLS) configuration files. The
.BR zipl.conf (5)
configuration file is still used to specify parameters or aditional boot configurations.
.PP

To use a single boot configuration section, provide its name as parameter to
.BR zipl .
If 
.B zipl
is called without parameters, it searches the configuration file for a
section specifying the default action.

.B Boot menu

The
.B zipl
tool implements a boot menu which includes the following features:
.IP "     -"
display a list of available configurations
.IP "     -"
choose a configuration
.IP "     -"
specify additional kernel command line parameters
.PP

See the
.BR zipl.conf (5)
man page for details on how to use the boot menu.

.B Logical devices

zipl can be used to prepare logical devices (e.g. a linear device mapper target)
for booting when the following requirements are met by the logical device setup:
.IP "     -"
all boot relevant files (i.e. kernel, ramdisk and parameter files) must be
located on a logical device which is mapped to a single physical disk of a type
supported by zipl (i.e. DASD or SCSI disk)
.IP "     -"
adjacent data blocks on the logical device must correspond to adjacent blocks on
the physical device
.IP "     -"
access to the first blocks (starting at block 0) of the physical device must be
given
.PP
Examples for logical device setups that are supported are linear and mirror
mapping.

When working with logical devices, zipl requires that the user provides more
information about the target device:
.IP "     -"
device characteristics of the underlying physical device: disk type and format
(e.g. ECKD CDL or FCP SCSI), disk geometry in case of ECKD DASDs and block size
.IP "     -"
target device offset, i.e. the number of blocks between the physical device
start and the start of the logical device containing the filesystem with all
boot relevant files
.IP "     -"
a device node which provides access to the first blocks of the device
.PP
If the user does not provide this information explicitly by parameters
zipl automatically runs a driver specific helper script to obtain these data,
e.g. zipl_helper.device-mapper.

Note that zipl uses /proc/devices to determine the driver name for a given
device. If the driver name cannot be determined the preparation of a logical
device for boot might fail.
This can be the case in a chroot environment when /proc is not mounted
explicitly.

.SH OPTIONS
.TP
.BR "\-h" " or " "\-\-help"
Print usage information, then exit.

.TP
.BR "\-v" " or " "\-\-version"
Print version information, then exit.

.TP
.BR "\-c <CONFIG FILE>" " or " "\-\-config=<CONFIG FILE>"
Use the specified <CONFIG FILE>. If none is supplied, the environment
variable ZIPLCONF is evaluated if set, otherwise /etc/zipl.conf is used.

.TP
.BR "\-b <BLS DIRECTORY>" " or " "\-\-blsdir=<BLS DIRECTORY>"
Use the specified <BLS DIRECTORY> to parse BootLoaderSpec config files.
If none is supplied, the /boot/loader/entries directory is used.

.TP
.BR "\-t <TARGET DIRECTORY>" " or " "\-\-target=<TARGET DIRECTORY>"
Use the specified <TARGET DIRECTORY>. 
.B zipl
uses this directory to store the bootmap, i.e. a file containing
boot data. The actual boot loader is installed onto the device containing
the target directory. Supported devices are DASD and SCSI disks.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-\-targetbase=<BASE DEVICE>"
Install the actual boot loader on the device node specified by BASE DEVICE.

This option is required when working with logical devices (see section
"Logical devices" above).

.TP
.BR "\-\-targettype=<TARGET TYPE>"
Assume that the physical device is of the specified type. Valid values are:
.IP "         -" 12
CDL: DASD disk with ECKD/compatible disk layout
.IP "         -" 12
LDL: DASD disk with ECKD/linux disk layout
.IP "         -" 12
FBA: FBA disk DASD
.IP "         -" 12
SCSI: SCSI disk
.PP
.IP " " 8
This option is required when working with logical devices (see section
"Logical devices" above).

.TP
.BR "\-\-targetgeometry=<CYLINDERS,HEADS,SECTORS>"
Assume that the physical device has the specified number of cylinders, heads and
sectors.

This option is required when working with logical devices which are located on
DASD ECKD disks (see section "Logical devices" above).

.TP
.BR "\-\-targetblocksize=<SIZE>"
Assume that blocks on the physical device are SIZE bytes long.

This option is required when working with logical devices (see section
"Logical devices" above).

.TP
.BR "\-\-targetoffset=<OFFSET>"
Assume that the logical device containing the directory specified by the
--target option is located on the physical device starting at the block
specified by OFFSET.

This option is required when working with logical devices (see section
"Logical devices" above).

.TP
.BR "\-T <TAPE DEVICE>" " or " "\-\-tape=<TAPE DEVICE>"
Install bootloader on the specified <TAPE DEVICE>. Use this option instead
of the 'target' option to prepare a tape device for IPL.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-i <IMAGE[,ADDRESS]>" " or " "\-\-image=<IMAGE[,ADDRESS]>"
Use the Linux kernel image file <IMAGE>. An optional hexadecimal
ADDRESS may be provided to specify a non-standard load address for the
image file.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-r <RAMDISK[,ADDRESS]>" " or " "\-\-ramdisk=<RAMDISK[,ADDRESS]>"
Use the ramdisk image <RAMDISK>. An optional hexadecimal
ADDRESS may be provided to specify a non-standard load address for the
ramdisk file.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-p <PARMFILE[,ADDRESS]>" " or " "\-\-parmfile=<PARMFILE[,ADDRESS]>"
Use the kernel command line stored in file <PARMFILE>. An optional hexadecimal
ADDRESS may be provided to specify a non-standard load address for the
parameter file.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-P <PARMLINE>" " or " "\-\-parameters=<PARMLINE>"
When installing a kernel, use the kernel command line <PARMLINE>.
When --parmfile is specified as well, the contents of the parmfile
and <PARMLINE> are appended to form the resulting command line.

It is not possible to specify both this parameter and the name of a menu
section on the command line at the same time.

.TP
.BR "\-k auto" " or " "\-\-kdump=auto"
Install a kdump kernel that can be used as a stand-alone dump tool. You
can IPL this kernel in an LPAR or guest virtual machine to create a dump of
a previously running operating system instance that has been configured
with a reserved memory area for kdump. For Linux, the memory is reserved
with the "crashkernel" kernel parameter.

.TP
.BR "\-s <SEGMENT,ADDRESS>" " or " "\-\-segment=<SEGMENT,ADDRESS>"
Use the segment image <SEGMENT>. The mandatory hexadecimal ADDRESS specifies
the load address for the segment file.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-d <DEVNODE[,SIZE]>" " or " "--dumpto=<DEVNODE[,SIZE]>"
Install a system dump record on the device identified by DEVNODE.
Supported devices are DASD ECKD or FBA disk partitions,
device mapper multipath partitions of FCP attached SCSI disks and IBM
3480/3490/3590/3592 tape devices.

With the exception of SCSI, an optional decimal SIZE parameter may be
specified to determine the maximum dump size in bytes. SIZE can be suffixed
by either of the letters K, M or G to signify that the decimal number be
interpreted as kilobytes, megabytes or gigabytes respectively. SIZE will be
rounded up to the next megabyte boundary. Note that when you specify a SIZE
lower than the actual memory size used by linux (see kernel parameter mem=),
the resulting dump will be incomplete.

Note that on partitions formatted with the ECKD/linux disk layout, the dump
record will be overwritten by system dump data so that it can be used for
IPL only once before the system dump record has to be installed again.
Partitions formatted with the ECKD/compatible disk layout allow for an
arbitrary number of subsequent dumps.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-D" " or " "\-\-dumptofs"
This option has been removed, use --dumpto instead.

.TP
.BR "\-M <DUMPLIST[,SIZE]>" " or " "--mvdump=<DUMPLIST[,SIZE]>"
Install a multi-volume dump record on each device associated with one of the
partitions listed in file DUMPLIST.
Supported are DASD ECKD partitions formatted with the compatible
disk layout.

A dump signature is written to each partition contained in
DUMPLIST. This signature is checked by the dump tool, when the dump is
written. If a dump partition does not have the signature, the dump tool
will refuse to write the dump. With the
.BR \-\-force
option this mechanism can be disabled. E.g. this can be used
for installing swap space on dump partitions.

An optional decimal SIZE parameter may be specified to determine the
maximum dump size in bytes. SIZE can be suffixed by either of the letters
K, M or G to signify that the decimal number be interpreted as kilobytes,
megabytes or gigabytes respectively. SIZE will be rounded up to the next
megabyte boundary. Note that when you specify a SIZE lower than the actual
memory size used by linux (see kernel parameter mem=), the resulting dump
will be incomplete.

This option is useful if the actual memory size used by linux is
larger than any single ECKD DASD partition could hold.

It is not possible to specify both this parameter and the name of a menu
or configuration section on the command line at the same time.

.TP
.BR "\-m <MENU>" " or " "--menu=<MENU>"
Install the multi-boot configuration defined in menu section <MENU> of the
configuration file.

This option cannot be used together with either 
.BR \-\-target ,
.BR \-\-image ,
.BR \-\-ramdisk ,
.BR \-\-parmfile ,
.BR \-\-parameters ,
.BR \-\-segment ,
.BR \-\-dumpto " or"
.BR \-\-mvdump .

.TP
.BR "\-n" " or " "\-\-noninteractive"
Answer all confirmation questions with 'yes'. This option may be useful when
.B zipl
is called from within a shell script or in other situations where no user
interaction is possible.

.TP
.BR "\-V" " or " "\-\-verbose"
Provide more verbose output.

.TP
.BR "\-a" " or " "\-\-add-files"
Copy all specified files to the bootmap file instead of just referencing them.
This option allows specifying files in a boot configuration which are not
located on the target device.

.TP
.B "\-\-dry\-run"
Print the results of performing the specified action without actually changing
the IPL records. This option can be used to test a configuration since all
error-checking is still done.

Note that when working with tapes, the current device position will be changed
even when specifying this option.

.TP
.BR "\-f" " or " "\-\-force"
Omit validity checking while producing a multi-volume dump.
Multi-volume dump partitions will be overwritten with dump data no matter
whether they contain a dump signature or not.

This option can only be used together with
.BR \-\-mvdump .

.TP
.BR "\-S <SWITCH>" " or " "\-\-secure <SWITCH>"
Control the zIPL secure boot support.
<SWITCH> can take one of three values:

  auto (default)
    Write signatures if available and supported by the system.
  1
    Write signatures regardless of support indicated by the local
    system. Missing signatures for stage 3 and kernel IPL files
    will result in an error.
  0
    Do not write signatures.


.SH EXAMPLE
1. Scenario: prepare disk for booting a Linux kernel image using the
following parameters:
.br
  - the directory /boot is located on the boot device
.br
  - /boot/image contains the Linux kernel image
.br
  - /boot/ramdisk.img contains a ramdisk image
.br
  - /boot/parmfile contains the kernel parameter line
.br

.RB "The respective " "zipl " "call reads:"
.br

  zipl -t /boot -i /boot/image -r /boot/ramdisk.img
.br
       -p /boot/parmfile
.br

2. Scenario: prepare DASD partition /dev/dasda1 for system dump:
.br

  zipl -d /dev/dasda1
.br

.SH NOTES
While it is not recommended for reasons of recovery and redundancy, FCP.
attached SCSI disks can also be accessed directly without multipathing,.
for example via the "/dev/disk/by-path/" device nodes.

.SH SEE ALSO
.BR zipl.conf (5),
.BR kexec (8)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------
 * __v2.9.x (2019-xx-xx)__

  For Linux kernel version: 5.xx

  Changes of existing tools:

  Bug Fixes:


 * __v2.9.0 (2019-05-21)__

  For Linux kernel version: 5.0 / 5.1

  Add new tool:
  - zcryptstats: Add zcryptstats to display usage statistics of
      IBM Crypto Express adapters

  Changes of existing tools:
  - lszfcp: New command line option to show module parameters
  - lszfcp: Sdev attributes for scsi_disk, block, integrity, queue, iosched
  - lszfcp: Add new output marker for non-good SCSI devices (luns)
  - lszfcp: Add new output marker for non-good fc_rports
  - lszfcp: Clean up whitespace (mixed indentation, trailing)
  - lschp: Add support for specifying a CHPID
  - zipl: Add secure boot capabilities
  - zkey: Add common passphrase options for cryptsetup and crypttab
  - zkey: Add batch-mode option to cryptsetup and zkey-cryptsetup
  - libu2s: Remove the entire library and provide more robust functionality
      in libdasd and libutil instead

  Bug Fixes:
  - lszfcp: Allow to show zfcp_units without associated SCSI device
  - lszfcp: Attribute details for: css, zfcp_port, zfcp_unit
  - lszfcp: Allow to also enumerate FCP device that have never been online
  - lszfcp: Fix error message if no zfcp-attached SCSI device found
  - lszfcp: Fix to show defunct FCP devices again
  - lszfcp: Fix to show non-good target ports again
  - lszfcp: Fix missing block & sg device output without CONFIG_SYSFS_DEPRECATED
  - lszfcp: New command line option for extended output format
  - zfcpdbf: Warn about ambiguous payload records with dup reqid & payarea
  - zpcictl: Check for regular directory to prevent possible buffer overflow

 * __v2.8.0 (2019-02-15)__

  For Linux kernel version: 4.20

  Changes of existing tools:
  - Switch to using /run directory instead of the legacy /var/run
  - zkey: Add --pbkdf pbkdf2 to generated cryptsetup luksFormat command
  - zdsfs: Add online VTOC refresh
  - pkey: Support autoloading kernel pkey module

  Bug Fixes:
  - zkey: Avoid EPERM on key change if user is not owner of key file
  - cpumf/cpumf_helper: Always return list reference for --sfb-size

 * __v2.7.1 (2018-12-13)__

  For Linux kernel version: 4.19

  Changes of existing tools:
  - zkey: Enhance file read/write error handling
  - cmsfs-fuse: Write more than a single line in linefeed mode
  - zpcictl: Add warning for unsupported operations
  - zipl: Use the BLS "title" field as the IPL section name

  Bug Fixes:
  - cmsfs-fuse: Fix iconv buffer aliasing
  - cmsfs-fuse: Fix memory leak in cmsfs_rename()
  - fdasd: Fix possible integer overflow
  - fdasd: Fix resource leak in fdasd_parse_conffile()
  - zdev: Fix memory leak in misc_readlink()
  - dasdinfo: Display error messages on stderr
  - zkey: Include /sbin into PATH when executing commands
  - Makefile: Fix parallel build
  - GCC8 warning fixes across the board for:
      cmsfs-fuse, dasdinfo, dasdview, dump2tar, fdasd, hmcdrvfs, hyptop,
      ip_watcher, libvmdump, libvtoc, lsqeth, lszcrypt, qethqoat, zdev, zdsfs,
      zgetdump, zipl, zpcictl

 * __v2.7.0 (2018-10-31)__

  For Linux kernel version: 4.19

  Add new tool:
  - zcryptctl: Add zcryptctl for multiple zcrypt node management
  - zpcictl: Add zpcictl for reporting defective PCI devices

  Changes of existing tools:
  - qethqoat: Add OSA-Express7S support
  - lszcrypt: Add support for alternative zcrypt device drivers
  - zfcpdump: Add install script for zfcpdump
  - zipl: Make zipl work with XFS by using the FIEMAP mapping ioctl

  Bug Fixes:
  - lstape: Fix output with SCSI lin_tape and multiple paths to same unit
  - lstape: Fix output without SCSI generic (sg)
  - lsluns: Fix to prevent error messages if there are non-zfcp SCSI devices
  - lstape: Fix to prevent error messages if there are non-zfcp SCSI devices
  - lstape: Fix description of --type and <devbusid> filter for channel tapes
  - lstape: Fix SCSI output description in man page
  - lstape: Fix SCSI HBA CCW device bus-ID e.g. for virtio-scsi-ccw
  - Direct --help and --version output to stdout for several tools
  - osasnmpd: Start without real OSA devices

 * __v2.6.0 (2018-08-10)__

  For Linux kernel version: 4.18

  Add new tool:
  - zkey: Add zkey-cryptsetup tool
  Changes of existing tools:
  - netboot: add BOOTIF support

  Bug Fixes:
  - mon_procd: fix parsing of /proc/<pid>/stat
  - netboot: Include compressed kernel modules in initramfs
  - netboot: Send client architecture and handle path prefix

 * __v2.5.0 (2018-06-08)__

  For Linux kernel version: 4.17

  Changes of existing tools:
  - zdev: Add support for reading firmware configuration files
  - zipl: Add BootLoaderSpec support
  - scripts: Add script to switch zipl config to a BootLoaderSpec setup

  Bug Fixes:
  - lsluns: Print a message if no adapter or port exists

 * __v2.4.0 (2018-05-07)__

  For Linux kernel version: 4.16

  Changes of existing tools:
  - dbginfo: Gather nvme related data
  - zipl: Rewrite helper script in C
  - libutil: Add function util_strstrip
  - libvmcp: Introduce libvmcp
  - zipl: Extend DASD stand-alone dumpers to drop zero pages
  - zgetdump: Add verbose option
  - zgetdump: Add 'Dump file size' field for zgetdump -i output
  - cpumf: Add IBM z14 ZR1 to the CPU Measurement Facility model list
  - zkey: Add build dependency to OpenSSL (libcrypto)
  - zkey: Add keystore implementation


  Bug Fixes:
  - hmcdrvfs: fix parsing of link count >= 1000
  - zgetdump: Avoid Segfault on processing dumps with memory limit
  - chreipl: correct fcp reipl sysfs write sequence
  - udev: Replace WAIT_FOR with TEST keyword

* __v2.3.0 (2018-01-30)__

  For Linux kernel version: 4.15

  Changes of existing tools:
  - lscpumf: Add support for IBM z14 hardware counters
  - libdasd: Introduce libdasd and use for dasd tools
  - zipl: Always build and link without PIE

  Bug Fixes:

zipl/src/zipl.c
/*
 * zipl - zSeries Initial Program Loader tool
 *
 * Main function
 *
 * Copyright IBM Corp. 2001, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include "lib/zt_common.h"

#include "boot.h"
#include "bootmap.h"
#include "disk.h"
#include "error.h"
#include "install.h"
#include "job.h"
#include "misc.h"
#include "zipl.h"


/* Flag deciding the level of verbosity */
int verbose = 0;

/* Flag deciding whether confirmation questions are asked */
int interactive = 1;

/* Flag deciding whether actions should only be simulated */
int dry_run = 1;

/* Full tool name */
static const char tool_name[] = "zipl: zSeries Initial Program Loader";

/* Copyright notice */
static const char copyright_notice[] = "Copyright IBM Corp. 2001, 2017";

/* Usage information */
static const char* usage_text[] = {
"Usage: zipl [OPTIONS] [SECTION]",
"",
"Prepare a device for initial program load. Use OPTIONS described below or ",
"provide the name of a SECTION defined in the zIPL configuration file.",
"",
"-h, --help                      Print this help, then exit",
"-v, --version                   Print version information, then exit",
"-c, --config CONFIGFILE         Read configuration from CONFIGFILE",
"-b, --blsdir BLSDIR             Parse BootLoaderSpec files from BLSDIR",
"-t, --target TARGETDIR          Write bootmap file to TARGETDIR and install",
"                                bootloader on device containing TARGETDIR",
"    --targetbase BASEDEVICE     Install bootloader on BASEDEVICE",
"    --targettype TYPE           Use device type: CDL, LDL, FBA, SCSI",
"    --targetgeometry C,H,S      Use disk geometry: cylinders,heads,sectors",
"    --targetblocksize SIZE      Use number of bytes per block",
"    --targetoffset OFFSET       Use offset between logical and physical disk",
"-i, --image IMAGEFILE[,ADDR]    Install Linux kernel image from IMAGEFILE",
"-r, --ramdisk RAMDISK[,ADDR]    Install initial ramdisk from file RAMDISK",
"-p, --parmfile PARMFILE[,ADDR]  Use kernel parmline stored in PARMFILE",
"-P, --parameters PARMLINE       Use specified kernel PARMLINE",
"-T, --tape TAPEDEV              Install bootloader on tape device TAPEDEV",
"-s, --segment SEGMENT,ADDR      Install a segment from file SEGMENT",
"-d, --dumpto DUMPDEV[,SIZE]     Install a system dump record on tape device",
"                                or disk partition DUMPDEV",
"-M, --mvdump DEVLIST[,SIZE]     Install a multi-volume dump record on each",
"                                disk partition listed in file DEVLIST",
"-f, --force                     Disable sanity check while producing a",
"                                multi-volume dump",
"-m, --menu MENU                 Install multi-boot configuration MENU",
"-n, --noninteractive            Answer all confirmation questions with 'yes'",
"-V, --verbose                   Provide more verbose output",
"-a, --add-files                 Add all referenced files to bootmap file",
"    --dry-run                   Simulate run but don't modify IPL records"
};


/* Print usage information. */
static void
print_usage(void)
{
	unsigned int i;

	for (i=0; i < ARRAY_SIZE(usage_text); i++)
		printf("%s\n", usage_text[i]);
}


/* Print version information. */
static void
print_version(void)
{
	printf("%s version %s\n", tool_name, RELEASE_STRING);
	printf("%s\n", copyright_notice);
}


/* Check whether calling user is root. Return 0 if user is root, non-zero
 * otherwise. */
static int
check_for_root(void)
{
	if (geteuid() != 0) {
		error_clear_text();
		error_reason("Must be root to perform this operation");
		return -1;
	} else
		return 0;
}


int
main(int argc, char* argv[])
{
	struct disk_info* info;
	disk_blockptr_t program_table, scsi_dump_sb_blockptr, *stage1b_list;
	blocknum_t stage1b_count;
	struct job_data* job;
	char* device;
	int rc;

	/* Check internals */
	rc = boot_check_data();
	if (rc) {
		error_text("Internal error");
		error_print();
		return 1;
	}
	/* Find out what we're supposed to do */
	rc = job_get(argc, argv, &job);
	if (rc) {
		error_print();
		return 1;
	}
	/* Check for priority options --help and --version */
	if (job->id == job_print_usage) {
		print_usage();
		job_free(job);
		return 0;
	} else if (job->id == job_print_version) {
		print_version();
		job_free(job);
		return 0;
	}
	/* Make sure we're running as root */
	if (check_for_root()) {
		job_free(job);
		error_print();
		return 1;
	}
	/* Set global option variables */
	verbose = job->verbose;
	interactive = !job->noninteractive;
	dry_run = job->dry_run;
	if (dry_run)
		printf("Starting dry-run, target device contents will NOT be "
		       "modified\n");
	/* Make sure new files are only user-accessible */
	umask(077);
	/* Do it */
	switch (job->id) {
	case job_dump_partition:
		if (disk_is_tape(job->data.dump.device) ||
		    !disk_is_scsi(job->data.dump.device, &job->target)) {
			rc = install_dump(job->data.dump.device, &job->target,
					  job->data.dump.mem);
			break;
		}
		/* Dump to a raw SCSI partition */
		if (job->data.dump.mem != -1uLL) {
			error_reason("Dump size can not be limited for "
				     "partition dump on a SCSI disk");
			rc = -1;
			break;
		}
		rc = check_job_dump_images(&job->data.dump, job->name);
		if (rc != 0)
			break;
		/* Fall through. */
	case job_ipl:
	case job_segment:
	case job_menu:
		/* Create bootmap */
		stage1b_list = NULL;
		rc = bootmap_create(job, &program_table, &scsi_dump_sb_blockptr,
				    &stage1b_list, &stage1b_count, &device,
				    &info);
		if (rc)
			break;
		/* Install boot loader */
		rc = install_bootloader(device, &program_table,
					&scsi_dump_sb_blockptr, stage1b_list,
					stage1b_count, info, job);
		if (stage1b_list != NULL)
			free(stage1b_list);
		misc_free_temp_dev(device);
		disk_free_info(info);
		break;
	case job_ipl_tape:
		rc = install_tapeloader(job->data.ipl_tape.device,
					job->data.ipl_tape.image,
					job->data.ipl_tape.parmline,
					job->data.ipl_tape.ramdisk,
					job->data.ipl_tape.image_addr,
					job->data.ipl_tape.parm_addr,
					job->data.ipl_tape.ramdisk_addr);
		break;
	case job_mvdump:
		rc = install_mvdump(job->data.mvdump.device,
				    &job->target,
				    job->data.mvdump.device_count,
				    job->data.mvdump.mem,
				    job->data.mvdump.force);
		break;
	case job_print_usage:
	case job_print_version:
		/* Should not happen */
		break;
	}
	switch (rc) {
	case 0: /* Operation completed successfully */
		if (verbose)
			printf("Syncing disks...\n");
		if (!dry_run)
			sync();
		printf("Done.\n");
		break;
	case -2: /* Operation canceled by user */
		break;
	default: /* An error occurred */
		error_print();
		break;
	}
	job_free(job);
	return abs(rc);
}
